tools/checklocks/state.go

```diff
@@ -18,23 +18,52 @@ import (
 	"fmt"
 	"go/token"
 	"go/types"
+	"sort"
 	"strings"
 	"sync/atomic"
-	"sort"
 
 	"golang.org/x/tools/go/ssa"
 )
 
 // lockType represents the current state of the lock.
 // The 0 value represents the unlocked state.
+// The valid states during analysis are:
+// unlocked, locked, deferredunlock and locked & deferredunlock
 type lockType int
 
 const (
-	unlocked lockType = iota
-	locked
-	// TODO(jamesyou): DEFERUNLOCK
+	unlocked lockType = 0
+	locked   lockType = 1 << iota
+	deferredunlock
 )
 
+// is checks if the lockType is in the state `t`.
+func (l lockType) is(t lockType) bool {
+	if t == unlocked {
+		return l&locked == 0
+	} else {
+		return l&t > 0
+	}
+}
+
+// union returns the union of the two lockTypes.
+// unlocked is the identity value, `l` union unlocked == `l`
```

Comment:
"identity value" kinda comes out of the blue. I think this won't make sense without more documentation of how readers should think about lockType.